bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -482,16 +490,33 @@ public PlacementResult<BookieId> replaceBookie(int ensembleSize, int writeQuorum
 
             if (LOG.isDebugEnabled()) {
                 LOG.debug("Try to choose a new bookie to replace {} from ensemble {}, excluding {}.",
-                    bookieToReplace, ensembleNodes, excludeNodes);
+                        bookieToReplace, ensembleNodes, excludeNodes);
             }
             // pick a candidate from same rack to replace
-            BookieNode candidate = selectFromNetworkLocation(
-                    bn.getNetworkLocation(),
-                    networkLocationsToBeExcluded,
-                    excludeNodes,
-                    TruePredicate.INSTANCE,
-                    EnsembleForReplacementWithNoConstraints.INSTANCE,
-                    !enforceMinNumRacksPerWriteQuorum);
+            BookieNode candidate;
+            try {
+                candidate = selectFromNetworkLocation(
+                        bn.getNetworkLocation(),
+                        networkLocationsToBeExcluded,
+                        excludeNodes,
+                        TruePredicate.INSTANCE,
+                        EnsembleForReplacementWithNoConstraints.INSTANCE,
+                        !enforceMinNumRacksPerWriteQuorum);
+            } catch (BKNotEnoughBookiesException e) {
+                if (downgradeToSelf) {
+                    candidate = knownBookies.get(bookieToReplace);
```

Comment:
We need to ensure the bookie is writable. The knownBookies set contains all the bookies which are writable and read-only.

Reply:
Please cover the read-only bookies in the unit test.

Reply:
> We need to ensure the bookie is writable. The knownBookies set contains all the bookies which are writable and read-only.

No, the knownBookies only contain writable bookies. The readOnlyBookies will store the read-only bookies.